File: Programming and Projection Methods/Lab7/Laborator7/CompanyClient/src/main/java/company/client/Controller.java
package company.client;

import company.model.Client;
import company.model.RBooking;

import company.model.Ride;
import company.model.User;
import company.services.IObserver;
import company.services.IServer;
import javafx.application.Platform;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.util.Pair;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;


public class Controller implements IObserver {
    private IServer server;
    private User user;
    private ObservableList<Ride> rModel;
    private List<Pair<ObservableList<RBooking>, List<String>>> rbModelList;

    public Controller(IServer serv) {
        this.server = serv;
        rModel = FXCollections.observableArrayList();
        rbModelList = new ArrayList<>();
    }

    private void populateRideList() {
        rModel.clear();
        Ride[] rides = server.findAllRides();
        for (Ride r : rides) {
            rModel.add(r);
        }
    }

    public ObservableList<Ride> getRModel() {
        populateRideList();
        return rModel;
    }

    public ObservableList<RBooking> getRBModel(String destination, String date, String hour) {
        ObservableList<RBooking> rbModel = FXCollections.observableArrayList();
        RBooking[] rbs = server.findAllRBookings(destination, date, hour);
        for (RBooking rb : rbs) {
            rbModel.add(rb);
        }
        //forEach(x -> rbModel.add(x));
        rbModelList.add(new Pair<>(rbModel, Arrays.asList(destination, date, hour)));
        return rbModel;
    }

    public void findUser(String username, String password) {
        User user = new User(username, password);
        server.login(user, this);
        this.user = user;
    }

    public void logout(User u) {
        server.logout(u, this);
    }

    public String bookPlaces(Ride r, String cname, int nrplaces) {
        return server.bookPlaces(r, new Client(0, cname), nrplaces);
    }

    private void upd(Pair<ObservableList<RBooking>, List<String>> pair) {
        pair.getKey().clear();
        RBooking[] rbs = server.findAllRBookings(pair.getValue().get(0), pair.getValue().get(1), pair.getValue().get(2));
        for (RBooking rb : rbs) {
            pair.getKey().add(rb);
        }
    }

    @Override
    public void update() {
        Platform.runLater(new Runnable() {
            @Override
            public void run() {
                populateRideList();
                rbModelList.forEach(x -> upd(x));
            }
        });
    }
}

File: Programming and Projection Methods/Lab8C#/Laborator8/CompanyClient/Controller.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using CompanyServices;
using CompanyModel;

namespace CompanyClient
{
    public class Controller : IObserver
    {
        public event EventHandler<UserEventArgs> UpdateEvent;
        private IServer server;
        private User user;
        private IList<Object> rModel;
        private IList<KeyValuePair<IList<RBooking>, IList<String>>> rbModelList;

        public Controller(IServer server)
        {
            this.server = server;
            rModel = new List<Object>();
            rbModelList = new List<KeyValuePair<IList<RBooking>, IList<String>>>();
        }

        private void PopulateRideList()
        {
            rModel.Clear();
            Ride[] rides = server.FindAllRides();
            foreach (var r in rides)
                rModel.Add(new { Destinatia = r.Destination, Data = r.Date, Ora = r.Hour, NrLocuriDisponibile = r.NrPlacesAvailable });
        }

        public IList<Object> GetRModel()
        {
            PopulateRideList();
            return rModel;
        }

        public IList<RBooking> GetRBModel(string destination, string date, string hour)
        {
            IList<RBooking> rbModel = new ObservableCollection<RBooking>();
            server.FindAllRBookings(destination, date, hour).ToList().ForEach(x => rbModel.Add(x));
            rbModelList.Add(new KeyValuePair<IList<RBooking>, IList<string>>(rbModel, new List<string> { destination, date, hour }));
            return rbModel;
        }

        public User FindUser(String username, String password)
        {
            User user = new User(username, password);
            server.Login(user, this);
            this.user = user;
            return user;
        }

        public void Logout(User user)
        {
            server.Logout(user, this);
        }

        public String BookPlaces(Ride r, String cname, int nrplaces)
        {
            String places = server.BookPlaces(r, new Client(0, cname), nrplaces);
            return places;
        }

        private void Upd(KeyValuePair<IList<RBooking>, IList<string>> pair)
        {
            pair.Key.Clear();
            server.FindAllRBookings(pair.Value.ElementAt(0), pair.Value.ElementAt(1), pair.Value.ElementAt(2))
                .ToList().ForEach(x => pair.Key.Add(x));
        }

        protected virtual void OnUserEvent(UserEventArgs e)
        {
            if (UpdateEvent == null)
                return;
            UpdateEvent(this, e);
            Console.WriteLine("Update event called");
        }

        public void Update()
        {
            Console.WriteLine("Controller update");
            UserEventArgs userArgs = new UserEventArgs();
            OnUserEvent(userArgs);
        }
    }
}
